### DISCUSSION TITLE
Documentation with recursive nested dynamic pydantic models

### DISCUSSION THREAD

**Discussion by HappyDias** [fastapi/fastapi#7885] [Questions] [answered]
### First check

* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description

Is it possible to have runtime automated documentation created with nested pydantic models in fastapi endpoints? It seems that this works when the level of nested models is 1, when you increase the nested level, a KeyError occurrs when I try to access the documentation

### Example code:

```
import uvicorn
import asyncio

from typing import List

from fastapi import FastAPI

from pydantic import BaseModel, create_model

def make_model(more=0):
	values = {"message": (str, None)}
	class_name = "DynamicModelOuter" if more==0 else "DynamicModelInner"
	if more <= 1:
		values[f"messages{more}"] = (List[make_model(more=more+1)], [])

	model = create_model(class_name, **values)

	return model

def setup_app(app):
	NewModel = make_model()

	testinst = NewModel(**{
		"message": "Hello World test",
		"messages0":[
			{
				"message": "test",
				"messages1": [
					{
						"message":"test"
					}
				]
			}
		]
	})

	@app.get("/dynamic", response_model=NewModel, tags=["test"])
	async def dynamic():
		return {
			"message": "Hello World dynamic",
			"messages0":[
				{
					"message": "dynamic",
					"messages1":[
						{
							"message": "dynamic",
						}
					]
				}
			]
		}

	return app

if __name__ == "__main__":
	app = FastAPI()
	setup_app(app)
	uvicorn.run(app, host="127.0.0.1", port=8000, log_level="info")
```
This code should work as is, as long as you have pydantic, fastapi and uvicorn installed. Running this script launches a uvicorn server on port 8000 serving the fastAPI app object. You can access the `http://localhost:8000/dynamic` endpoint to see the result. However, when you access the docs endpoint `http://localhost:8000/docs` you get the following eye-gouging error (Windows output):
```
ERROR: Exception in ASGI application
Traceback (most recent call last):
  File "C:\Users\Antonio\AppData\Local\Programs\Python\Python37\lib\site-packages\uvicorn\protocols\http\h11_impl.py", line 375, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "C:\Users\Antonio\AppData\Local\Programs\Python\Python37\lib\site-packages\fastapi\applications.py", line 140, in __call__
    await super().__call__(scope, receive, send)
  File "C:\Users\Antonio\AppData\Local\Programs\Python\Python37\lib\site-packages\starlette\applications.py", line 134, in __call__
    await self.error_middleware(scope, receive, send)
  File "C:\Users\Antonio\AppData\Local\Programs\Python\Python37\lib\site-packages\starlette\middleware\errors.py", line 178, in __call__
    raise exc from None
  File "C:\Users\Antonio\AppData\Local\Programs\Python\Python37\lib\site-packages\starlette\middleware\errors.py", line 156, in __call__
    await self.app(scope, receive, _send)
  File "C:\Users\Antonio\AppData\Local\Programs\Python\Python37\lib\site-packages\starlette\exceptions.py", line 73, in __call__
    raise exc from None
  File "C:\Users\Antonio\AppData\Local\Programs\Python\Python37\lib\site-packages\starlette\exceptions.py", line 62, in __call__
    await self.app(scope, receive, sender)
  File "C:\Users\Antonio\AppData\Local\Programs\Python\Python37\lib\site-packages\starlette\routing.py", line 590, in __call__
    await route(scope, receive, send)
  File "C:\Users\Antonio\AppData\Local\Programs\Python\Python37\lib\site-packages\starlette\routing.py", line 208, in __call__
    await self.app(scope, receive, send)
  File "C:\Users\Antonio\AppData\Local\Programs\Python\Python37\lib\site-packages\starlette\routing.py", line 41, in app
    response = await func(request)
  File "C:\Users\Antonio\AppData\Local\Programs\Python\Python37\lib\site-packages\fastapi\applications.py", line 97, in openapi
    return JSONResponse(self.openapi())
  File "C:\Users\Antonio\AppData\Local\Programs\Python\Python37\lib\site-packages\fastapi\applications.py", line 89, in openapi
    openapi_prefix=self.openapi_prefix,
  File "C:\Users\Antonio\AppData\Local\Programs\Python\Python37\lib\site-packages\fastapi\openapi\utils.py", line 281, in get_openapi
    flat_models=flat_models, model_name_map=model_name_map
  File "C:\Users\Antonio\AppData\Local\Programs\Python\Python37\lib\site-packages\fastapi\utils.py", line 79, in get_model_definitions
    model_name = model_name_map[model]
KeyError: <class 'DynamicModelInner'>
```
However, the automated documentation works swimmingly if you change the limitation on line 3 of the `make_model` function from `if more <= 1:` to `if more < 1:` limiting the nested level of the dynamic models to 1.

I don't want to mark this as a bug, I'll leave it to you guys here to decide whether it is one or not after you get to try the above code. Hopefully you'll have no issues getting it to work with higher levels of recursion and I'm just doing something wrong.

###Additional info

I am running the following specs:
- Windows 10
- Python version 3.7.4
- fastapi version 0.44.1
- pydantic version 1.3? (I'm accessing pydantic.version.VERSION for this number)
- uvicorn version 0.9.0

Thanks a lot in advance for any information you can give me. This is actually my first issue posting ever on gh so please let me know if I'm doing anything wrong with the formatting.

**Comment by HappyDias** [comment]
Bumping (if that's even a thing).

Any further info on this issue you guys can provide?

**Comment by tiangolo** [comment, marked as answer]
@HappyDias I'm not sure I understand what you want to achieve. I suspect it can be simplified or designed differently to avoid complexity.

But anyway, here's a version of what I think you want to have that works:

```Python
from typing import List

from fastapi import FastAPI
from pydantic import create_model


def make_model(more=0, class_name="DynamicModelOuter"):
    values = {"message": (str, None)}
    if more <= 1:
        inner_class_name = f"DynamicModelInner{more+1}"
        values[f"messages{more}"] = (
            List[make_model(more=more + 1, class_name=inner_class_name)],
            [],
        )

    model = create_model(class_name, **values)

    return model


def setup_app(app):
    NewModel = make_model()

    @app.get("/dynamic", response_model=NewModel, tags=["test"])
    async def dynamic():
        return {
            "message": "Hello World dynamic",
            "messages0": [
                {"message": "dynamic", "messages1": [{"message": "dynamic",}]}
            ],
        }

    return app
```

**Comment by HappyDias** [comment]
@tiangolo , thanks a lot for the reply. I've run your code and it seems I have found out where I went wrong.

My code was simply a poor attempt at making a minimal working example of what my problem is. I am designing a server application that creates a bunch of CRUD endpoints for classes defined via ORM (using sqlalchemy). A bit like a class-based-views, but for which the interaction endpoints are automatically generated via fastAPI with the input and output models being created via pydantic to enrich the error handling and documentation. The idea was that via recursive functions I would be able to use the mapping created by the ORM to create models to handle the database relationships (one-to-one, many-to-one, etc...)

It seems that the issue, however, is either pydantic related, because in your example, the fact that your code provides a unique name for each nested model seems to fix the issue. You are right in saying that it probably needs better designing, most likely to ensure the uniqueness of the model names. 

Looking at the pydantic code, however, it does seem that there is some sort of in-memory "hash table" for the classes and I would not expect that models using the same names would somehow "corrupt" this store but apparently it does (i.e: the `testinst` variable contains a fully instantiated dynamically-nested model using non-unique names and there are no complaints, the error only arises when the same model is used as a response_model for the fastAPI endpoint)

I will dig around a bit further during the weekend and try to fix my code by ensuring unique names for my models. If this does fix my code then I will close this issue and will consider myself instructed in good code practices.

Thanks

**Comment by HappyDias** [comment]
Closing as it was indeed bad design on my part, ensuring dynamically generated classes don't "repeat" and names are unique fixes my issue.

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue :+1:

**Comment by AndreiPashkin** [comment]
@tiangolo, encountered the same problem when generating models based on non FastAPI-specific definitions.

**Comment by tiangolo** [comment]
@AndreiPashkin if you're having a problem it's better to create a new issue, follow the template, and provide a minimal example reproducing your problem.
